Handle DDE in infer_size_impl

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163868
* __->__ #163822

hit this while running VLLM with unbacked for model Qwen/Qwen2-1.5B-Instruct

laithsakka committed:
handle DDE in infer_size_impl

[ghstack-poisoned]

laithsakka committed:
Update on "handle DDE in infer_size_impl"


hit this while running VLLM with unbacked for model Qwen/Qwen2-1.5B-Instruct


[ghstack-poisoned]

ColinPeppler commented on aten/src/ATen/InferSize.h:
```diff
@@ -64,7 +64,23 @@ inline void infer_size_impl(
+  if (infer_dim) {
+    // numel is the product of known sizes it has to be divisible by newsize.
+    if constexpr (std::is_same_v<NumelType, c10::SymInt>) {
+      TORCH_MAYBE_SYM_CHECK(
+          sym_gt(numel, 0).sym_and(sym_eq(numel % newsize, 0)),
+          "shape '",
+          shape,
+          "' is invalid for input");
+    } else {
+      TORCH_CHECK(
+          numel > 0 && numel % newsize == 0,
+          "shape '",
+          shape,
+          "' is invalid for input of size ",
+          numel);
+    }
```
noob question, can we just use TORCH_MAYBE_SYM_CHECK for both cases? or there's a performance hit so we need `if constexpr`?

laithsakka replied:
does not compile.

laithsakka replied:
sym_gt(numel, 0) return type would be bool for non symint path so we cant call sym_and on it.

laithsakka replied:
mm let me try to intorduce && in Symbool and see if use only operators might work that way

laithsakka replied:
did not work operator> specialize when second arg is int for SymInt,

laithsakka commented:
ah need to handle the case for ==0 will take a look tomorrow

laithsakka commented:
ok need to fix issues before re-asking for reviews.

laithsakka committed:
Update on "Handle DDE in infer_size_impl"


hit this while running VLLM with unbacked for model Qwen/Qwen2-1.5B-Instruct


[ghstack-poisoned]

laithsakka committed:
Update on "Handle DDE in infer_size_impl"


hit this while running VLLM with unbacked for model Qwen/Qwen2-1.5B-Instruct


[ghstack-poisoned]

laithsakka committed:
Update on "Handle DDE in infer_size_impl"


hit this while running VLLM with unbacked for model Qwen/Qwen2-1.5B-Instruct


[ghstack-poisoned]

laithsakka commented:
ok all tests pass, will fix lint later.

laithsakka committed:
Update on "Handle DDE in infer_size_impl"


hit this while running VLLM with unbacked for model Qwen/Qwen2-1.5B-Instruct


[ghstack-poisoned]

laithsakka committed:
Update on "Handle DDE in infer_size_impl"


hit this while running VLLM with unbacked for model Qwen/Qwen2-1.5B-Instruct


[ghstack-poisoned]

laithsakka commented:
> test?

done
